### PR TITLE
AFT: locking issues

### PR DESCRIPTION
**Pull request by ashamis**
part of #1601

I am fixing some of the locking issues, additional PRs will follow-up on this work

- In replicate raft.h:404, on the BFT backup we would go through the committed tx and put them on the ledger. This actually results in double adding ledger entries, raft.h:452, does this already
- There was a re-entrant lock issue if we ended up having to append the ViewChangeConfirmation message to the ledger, we solve this by release the lock for a short period of time. This is safe because there is no concurrency dependency by adding the view and propagating the ledger which would occur next.

### CODE REVIEW

File: src/consensus/aft/raft.h [hunk 1]
```diff
@@ -1176,8 +1179,16 @@ namespace aft
...
+              LOG_INFO_FMT(
+                "AAAAAA - i:{} - recv_append_entries - before:{}",
+                i,
+                state->last_idx);
               state->last_idx =
                 executor->commit_replayed_request(tx, request_tracker);
+              LOG_INFO_FMT(
+                "AAAAAA - i:{} - recv_append_entries - after:{}",
+                i,
+                state->last_idx);
```
Comment (jumaffre) [inline src/consensus/aft/raft.h:1191]:
```suggestion
              state->last_idx =
                executor->commit_replayed_request(tx, request_tracker);
```

File: src/consensus/aft/raft.h [hunk 2]
```diff
@@ -595,12 +591,19 @@ namespace aft
                 *vc, id(), new_view, seqno, node_count()) &&
             get_primary(new_view) == id())
           {
+            // We need to reobtain the lock when writing to the ledger so we
+            // need to release it at this time.
```
Comment (jumaffre) [inline src/consensus/aft/raft.h:595]: Just making sure I've understood this correctly: `append_new_view()` appends to the ledger in `write_view_change_confirmation()`, which writes to the ledger though KV transaction commit, then `replicate()`?

File: src/consensus/aft/raft.h (hunk 1 above)
Comment (ashamis) [reply src/consensus/aft/raft.h:1191]: thanks! i thought i deleted that......

File: src/consensus/aft/raft.h (hunk 2 above)
Comment (ashamis) [reply src/consensus/aft/raft.h:595]: yes

**Review by jumaffre — APPROVED** [review]

**Pull request merged**